### PR TITLE
bpo-30241: implement contextlib.AbstractAsyncContextManager

### PR DESCRIPTION
**Pull request by JelleZijlstra**
https://bugs.python.org/issue30241

### CODE REVIEW

File: Lib/contextlib.py [hunk 1]
```diff
@@ -31,6 +32,28 @@ def __subclasshook__(cls, C):
...
+    @abc.abstractmethod
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        """Raise any exception triggered within the runtime context."""
+        return None
+
+    @classmethod
+    def __subclasshook__(cls, C):
+        if cls is AbstractAsyncContextManager:
+            if (any("__aenter__" in B.__dict__ for B in C.__mro__) and
+                any("__aexit__" in B.__dict__ for B in C.__mro__)):
```
Comment (ilevkivskyi) [inline Lib/contextlib.py]: I understand you just copied this from above, but I think both should be updated to support the explicit "anti-registration" ``__enter__ = None`` pattern, like all classes in ``collections.abc``, see ``_check_methods`` function in ``_collections_abc.py``.

Comment (JelleZijlstra) [reply Lib/contextlib.py]: Maybe that should be a separate issue? `AbstractContextManager` should arguably be fixed even in 3.6, and this PR only targets 3.7. I can fix `AbstractAsyncContextManager` though.

Comment (ilevkivskyi) [reply Lib/contextlib.py]: Yes, a separate PR for 3.6 is needed. But I think it makes sense to already fix the async version here.

Comment (JelleZijlstra) [reply Lib/contextlib.py]: Filed http://bugs.python.org/issue30266 to track that.

**Review by ilevkivskyi — APPROVED** [review]

**Comment by gvanrossum** [comment]
Wait, isn't this a new feature that should go into 3.7?

**Comment by ilevkivskyi** [comment]
> Wait, isn't this a new feature that should go into 3.7?

Oh yes, sorry, this one is not urgent then.

**Comment by 1st1** [comment]
Please add a NEWS entry and I'll merge this in.

**Comment by JelleZijlstra** [comment]
Thanks, just pushed a NEWS entry.

File: Doc/library/contextlib.rst [hunk 2]
```diff
@@ -29,6 +29,15 @@ Functions and classes provided:
    .. versionadded:: 3.6
 
 
+.. class:: AbstractAsyncContextManager
+
+   An :term:`abstract base class` similar to
```
Comment (1st1) [inline Doc/library/contextlib.rst]: I'm wondering if we could just copy the paragraph from `AbstractContextManager` doc. "similar to" isn't easy to read here.

**Comment by Kentzo** [comment]
The PR https://github.com/python/cpython/pull/4790 adds AsyncExitStack subclass. It should inherit from `AbstractAsyncContextManager`.

**Comment by ilevkivskyi** [comment]
@JelleZijlstra Could you please fix the merge conflict and update the docs as @1st1 proposed? I would like to have this merged soon.

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
Thanks @JelleZijlstra!

**Comment by JelleZijlstra** [comment]
Thanks, and sorry for dropping the ball on making the docs change!

**Comment by 1st1** [comment]
NP ;)

**Comment by ilevkivskyi** [comment]
Thank you @JelleZijlstra  and @1st1 !

**Pull request merged**